path.go

```diff
@@ -1,127 +1,223 @@
 package there
 
 import (
+	"fmt"
 	"strings"
 )
 
-type Path struct {
-	parts      []pathPart
-	ignoreCase bool
-}
+type method uint8
+
+const (
+	methodGet method = iota
+	methodHead
+	methodPost
+	methodPut
+	methodPatch
+	methodDelete
+	methodConnect
+	methodOptions
+	methodTrace
+	methods
+)
 
-type pathPart struct {
-	value    string
-	variable bool
+func methodToInt(method string) method {
+	switch method {
+	case MethodHead:
+		return methodHead
+	case MethodPost:
+		return methodPost
+	case MethodPut:
+		return methodPut
+	case MethodPatch:
+		return methodPatch
+	case MethodDelete:
+		return methodDelete
+	case MethodConnect:
+		return methodConnect
+	case MethodOptions:
+		return methodOptions
+	case MethodTrace:
+		return methodTrace
+	default:
+		return methodGet
+	}
 }
 
-// ConstructPath returns the path to match.
-func ConstructPath(pathString string, ignoreCase bool) Path {
-	split := splitUrl(pathString)
-	parts := make([]pathPart, len(split))
-	for i, s := range split {
-		variable := false
-		const variablePrefix = ":"
-		if strings.HasPrefix(s, variablePrefix) {
-			s = s[len(variablePrefix):]
-			for _, part := range parts {
-				if part.variable && part.value == s {
-					panic(pathString + " has defined the route param \"" + s + "\" more than once")
-				}
-			}
-			variable = true
-		}
+func methodToString(method method) string {
+	switch method {
+	case methodHead:
+		return MethodHead
+	case methodPost:
+		return MethodPost
+	case methodPut:
+		return MethodPut
+	case methodPatch:
+		return MethodPatch
+	case methodDelete:
+		return MethodDelete
+	case methodConnect:
+		return MethodConnect
+	case methodOptions:
+		return MethodOptions
+	case methodTrace:
+		return MethodTrace
+	default:
+		return MethodGet
+	}
+}
 
-		parts[i] = pathPart{
-			value:    s,
-			variable: variable,
-		}
+type (
+	matcher struct {
+		root   *node
+		static map[string]*node
 	}
 
-	path := Path{
-		parts:      parts,
-		ignoreCase: ignoreCase,
+	node struct {
+		name        string
+		children    map[string]*node
+		paramNode   *node
+		handler     [methods]Endpoint
+		middlewares [methods][]Middleware
 	}
-	return path
-}
+)
 
-func (p Path) ToString() string {
-	path := "/"
-	for i, part := range p.parts {
-		if part.variable {
-			path += ":"
+func (m *matcher) ensureNodeExists(path string) (*node, error) {
+	isDynamic := len(path) != 0 && path[0] == ':'
```

Comment:
refactor isDynamic to general util function, and check if `:` is included anywhere, in case routes dont start with `:` (unreasonable, but better safe than sorry)